### PR TITLE
feat(pagination): adding shadow parts

### PR DESCRIPTION
**Pull request by bruno-amorim**
### Related Ticket(s)

[ADCMS-5336](https://jsw.ibm.com/browse/ADCMS-5336)

### Description

Adding shadow parts to the pagination component, so that it can be styled outside of the shadow tree.
Snapshots updated.

### CODE REVIEW

File: packages/carbon-web-components/src/components/pagination/pagination.ts [hunk 1]
```diff
@@ -345,8 +353,11 @@ class CDSPagination extends FocusMixin(HostListenerMixin(LitElement)) {
       .join(' ');
 
     return html`
-      <div class="${prefix}--pagination__left">
-        <label for="select" class="${prefix}--pagination__text"
+      <div part="left-container" class="${prefix}--pagination__left">
+        <label
+          part="pagination-text-left"
```
Comment (m4olivei) [inline packages/carbon-web-components/src/components/pagination/pagination.ts:358]: Lets drop the `pagination-` prefix from all the new part names in this component. It's redundant when you consider the selectors a parent page would be using, eg. `cds-pagination::part(pagination-text-left)`. 

Also, given there are a few elements that share the class `cds--pagination__text`, I think we could improve by using two part names, one generic so parent pages could target them all in one selector, and one specific, using a modifier. Using this one as an example, `part="text text--left"`.

File: packages/carbon-web-components/src/components/pagination/pagination.ts [hunk 2]
```diff
@@ -25,6 +25,14 @@ import { carbonElement as customElement } from '../../globals/decorators/carbon-
  * Pagination UI.
  *
  * @element cds-pagination
+ * @csspart left-container - The left container. Usage: `cds-pagination::part(left-container)`
+ * @csspart pagination-text-left - The left pagination text. Usage: `cds-pagination::part(pagination-text-left)`
+ * @csspart pagination-items-count - The total amount of items. Usage: `cds-pagination::part(pagination-items-count)`
+ * @csspart right-container - The right container. Usage: `cds-pagination::part(right-container)`
+ * @csspart pagination-right-label - The right label. Usage: `cds-pagination::part(pagination-right-label)`
+ * @csspart pagination-text-right - The left pagination text. Usage: `cds-pagination::part(pagination-text-right)`
```
Comment (m4olivei) [inline packages/carbon-web-components/src/components/pagination/pagination.ts:33]: Name and description mix 'right' and 'left'.

**Review by m4olivei — CHANGES_REQUESTED** [review]

File: packages/carbon-web-components/src/components/pagination/pagination.ts [hunk 3]
```diff
@@ -345,8 +353,11 @@ class CDSPagination extends FocusMixin(HostListenerMixin(LitElement)) {
...
-      <div class="${prefix}--pagination__left">
-        <label for="select" class="${prefix}--pagination__text"
+      <div part="left-container" class="${prefix}--pagination__left">
+        <label
+          part="pagination-text-left"
+          for="select"
+          class="${prefix}--pagination__text"
           ><slot name="label-text">${itemsPerPageText}</slot></label
         >
         <cds-select
```
Comment (m4olivei) [inline packages/carbon-web-components/src/components/pagination/pagination.ts:363]: Missing `part` attributes on `<cds-select>` and `<cds-select-item>` elements, both part of the shadow root for this component.

File: packages/carbon-web-components/src/components/pagination/pagination.ts [hunk 4]
```diff
@@ -360,17 +371,22 @@ class CDSPagination extends FocusMixin(HostListenerMixin(LitElement)) {
...
+        <label
+          part="pagination-right-label"
+          for="select"
+          class="${prefix}--label ${prefix}--visually-hidden">
           ${formatLabelText({ count: totalPages })}
         </label>
         ${pagesUnknown || !totalItems
           ? html`
               <span
+                part="pagination-text-right"
```
Comment (m4olivei) [inline packages/carbon-web-components/src/components/pagination/pagination.ts:389]: Just below here from lines 394-424, again, missing `part` attributes on `<cds-select>` and `<cds-select-item>` elements, both part of the shadow root for this component.

File: packages/carbon-web-components/src/components/pagination/pagination.ts [hunk 5]
```diff
@@ -409,12 +425,16 @@ class CDSPagination extends FocusMixin(HostListenerMixin(LitElement)) {
...
+                class="${prefix}--pagination__text"
                 >${formatSupplementalText({ count: totalPages })}</span
               >
             `}
 
-        <div class="${prefix}--pagination__control-buttons">
+        <div
+          part="controls-container"
+          class="${prefix}--pagination__control-buttons">
           <cds-button
```
Comment (m4olivei) [inline packages/carbon-web-components/src/components/pagination/pagination.ts:438]: Missing `part` attribute on `<cds-button>` components. Here and on line 447.

**Review by m4olivei — CHANGES_REQUESTED** [review]

**Comment by bruno-amorim** [comment]
Hi @m4olivei,
All your comments have been addressed. Can you please take a look again?

Thanks a lot for your help!

File: packages/carbon-web-components/src/components/pagination/pagination.ts [hunk 6]
```diff
@@ -25,6 +25,18 @@ import { carbonElement as customElement } from '../../globals/decorators/carbon-
  * Pagination UI.
  *
  * @element cds-pagination
+ * @csspart left-container - The left container. Usage: `cds-pagination::part(left-container)`
+ * @csspart text - The generic shadow part for all text elements. Usage: `cds-pagination::part(text)`
+ * @csspart text-left - The left pagination text. Usage: `cds-pagination::part(text-left)`
```
Comment (m4olivei) [inline packages/carbon-web-components/src/components/pagination/pagination.ts:30]: Sorry, I wasn't clear before. We've been using a convention inspired by [BEM](https://getbem.com/naming/#modifier), where the variant or modifier is named with a double-dash separator. For instance, see @marcelojcs 's recent change to the [code-snippet element](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/feat/v2-shadow-parts/packages/carbon-web-components/src/components/code-snippet/code-snippet.ts#L436). If you aren't familiar with BEM, that page is well worth the read. It'll also help make sense of many of the class names you see in the Carbon projects, as well as in AEM. 
```suggestion
 * @csspart text--left - The left pagination text. Usage: `cds-pagination::part(text--left)`
```

File: packages/carbon-web-components/src/components/pagination/pagination.ts [hunk 7]
```diff
@@ -25,6 +25,18 @@ import { carbonElement as customElement } from '../../globals/decorators/carbon-
...
+ * @csspart items-count - The total amount of items. Usage: `cds-pagination::part(items-count)`
+ * @csspart right-container - The right container. Usage: `cds-pagination::part(right-container)`
+ * @csspart right-label - The right label. Usage: `cds-pagination::part(right-label)`
+ * @csspart text-right - The right pagination text. Usage: `cds-pagination::part(text-right)`
```
Comment (m4olivei) [inline packages/carbon-web-components/src/components/pagination/pagination.ts:34]:
```suggestion
 * @csspart text--right - The right pagination text. Usage: `cds-pagination::part(text--right)`
```

File: packages/carbon-web-components/src/components/pagination/pagination.ts [hunk 8]
```diff
@@ -345,11 +357,15 @@ class CDSPagination extends FocusMixin(HostListenerMixin(LitElement)) {
       .join(' ');
 
     return html`
-      <div class="${prefix}--pagination__left">
-        <label for="select" class="${prefix}--pagination__text"
+      <div part="left-container" class="${prefix}--pagination__left">
+        <label
+          part="text text-left"
```
Comment (m4olivei) [inline packages/carbon-web-components/src/components/pagination/pagination.ts:362]:
```suggestion
          part="text text--left"
```

File: packages/carbon-web-components/src/components/pagination/pagination.ts [hunk 9]
```diff
@@ -360,22 +376,28 @@ class CDSPagination extends FocusMixin(HostListenerMixin(LitElement)) {
...
+        <label
+          part="right-label"
+          for="select"
+          class="${prefix}--label ${prefix}--visually-hidden">
           ${formatLabelText({ count: totalPages })}
         </label>
         ${pagesUnknown || !totalItems
           ? html`
               <span
+                part="text text-right"
```
Comment (m4olivei) [inline packages/carbon-web-components/src/components/pagination/pagination.ts:394]:
```suggestion
                part="text text--right"
```

**Review by m4olivei — CHANGES_REQUESTED** [review]

**Review by m4olivei — APPROVED** [review]

**Pull request merged**